### PR TITLE
Implement Reparameterized version of Gamma

### LINKED ISSUE #25: Implement Reparameterized version of Gamma

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #25 

**DO NOT MERGE.** This PR will be moved to the pytorch org after https://github.com/pytorch/pytorch/pull/3841 is merged.

This implements reparameterized gradient for `distributions.Gamma`. The gradient is implemented by directly approximating the reparameterized gradient function `dx/dalpha` following [Knowles (2015)](https://arxiv.org/pdf/1509.01631.pdf). The approximation is accurate to within 1% relative error for a wide range of alphas.

## Derivation

Note that if `x ~ Gamma(alpha, beta)` then `x / beta ~ Gamma(alpha, 1)`. Since division is already implemented in PyTorch, we can thus reduce our problem to computing a reparameterized gradient of a standard gamma `x ~ Gamma(alpha) = Gamma(alpha, 1)` wrt `alpha`.

This PR implements a function `standard_gamma_grad(x, alpha)` that directly approximates the reparameterized gradient defined (for any continuous univariate distribution) as
```
                d/dalpha cdf(x; alpha)     d/dalpha cdf(x; alpha)
dx / dalpha = - ---------------------- = - ----------------------
                  d/dx cdf(x; alpha)           pdf(x; alpha)
```
This definition is used in the unit tests in `tests/test_distributions.py`, which compute `d/dalpha cdf(x;alpha)` via finite difference of the `scipy.stats.gamma.cdf()` function.

The approximation is split into three regions:
- For `x < 0.001` we differentiate the power-law approximation from Knowles (2015)
  ```
  cdf(x; alpha)  \approx  x**alpha / (alpha * Gamma(alpha))
  standard_gamma_grad(x, alpha) = -x/alpha * (log(x) - 1/alpha - digamma(alpha))
  ```
  Until `digamma()` is implemented in PyTorch, we use a finite difference of `lgamma()`.
- For `alpha > 30` we use the approximation
  ```
  standard_gamma_grad(x, alpha) = sqrt(x/alpha)
  ```
- For intermediate x,alpha we use a rational function approximation
   ```
   standard_gamma_grad(x, alpha) = exp(PQ(log(x / alpha), log(alpha)))
   ```
   where `PQ(u,v)` is a rational function of order 2 in u and 3 in v. This was trained using least squares minimizing squared relative error on ~20000 samples drawn from
  ```
  alpha ~ log_uniform(1e-5, 1e2)
  x ~ Gamma(alpha)
  ```

For complete derivation, see this [Jupyter Notebook](https://github.com/fritzo/notebooks/blob/master/gamma-reparameterized.ipynb).

### CODE REVIEW

**Comment by fritzo** [comment]
@apaszke Any advice on autograd plumbing? (I'll send this PR to pytorch/pytorch after https://github.com/pytorch/pytorch/pull/3841 is merged)

File: torch/distributions.py [hunk 1]
```diff
@@ -197,6 +198,26 @@ def log_prob(self, value):
         return -((value - self.mean) ** 2) / (2 * var) - log_std - math.log(math.sqrt(2 * math.pi))
 
 
+class _StandardGamma(Function):
```
Comment (apaszke) [inline torch/distributions.py]: This is an old-style autograd function. You should write it as shown [in the docs](http://pytorch.org/docs/master/notes/extending.html#extending-torch-autograd).

File: aten/src/TH/THRandom.c [hunk 2]
```diff
@@ -305,6 +305,78 @@ double THRandom_standard_gamma(THGenerator *_generator, double alpha) {
   }
 }
 
+// This is identical to THRandom_standard_gamma but also stores the
+// reparameterized gradient wrt alpha in grad_alpha.
+double THRandom_standard_gamma_with_grad(THGenerator *_generator, double alpha,
```
Comment (apaszke) [inline aten/src/TH/THRandom.c]: You need to compute the grad during forward, because you'd need to replay all the control flow here otherwise, right?

File: torch/csrc/generic/methods/TensorRandom.cwrap [hunk 3]
```diff
@@ -228,6 +228,15 @@
           default: THPDefaultGenerator->cdata
           kwarg_only: True
         - THTensor* alpha
+    - cname: standard_gamma_alpha_with_grad
```
Comment (apaszke) [inline torch/csrc/generic/methods/TensorRandom.cwrap]: It's bette to expose this as an internal method instead of a new overload (think `torch._C._standard_gamma_alpha_with_grad`)

File: torch/autograd/variable.py [hunk 4]
```diff
@@ -332,6 +332,11 @@ def multinomial(self, num_samples=1, replacement=False):
     def bernoulli(self):
         return Variable(torch.bernoulli(self.data))
 
+    def standard_gamma(self, grad=None):
+        if grad is None:
+            return Variable(torch.standard_gamma(self.data))
+        return Variable(torch.standard_gamma(self.data, grad), requires_grad=self.requires_grad)
```
Comment (apaszke) [inline torch/autograd/variable.py]: You should just use the Function subclass you implemented here.

Comment (fritzo) [reply torch/autograd/variable.py]: Thanks for the help! Should I embed that subclass in `torch.autograd.variable`, or should `torch.distributions` be the canonical interface for `standard_gamma()` for `Variable`s?

File: torch/distributions.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions.py]: I wasn't able to store the gradient via `ctx.save_for_backward(grad)`. Is there a new-style way to save an intermediate that is neither an input nor an output?

File: torch/csrc/generic/methods/TensorRandom.cwrap (hunk 3 above)
Comment (fritzo) [reply torch/csrc/generic/methods/TensorRandom.cwrap]: Could you point me to an example internal method whose plumbing I can copy? I'm a little lost here.

File: torch/distributions.py (hunk 1 above)
Comment (apaszke) [reply torch/distributions.py]: You can't do that because that can only be done with inputs/outputs. Just do `ctx.grad = grad`.

File: aten/src/TH/THRandom.c (hunk 2 above)
Comment (fritzo) [reply aten/src/TH/THRandom.c]: Correct. This is a complicated gradient due to the rejection sampler. To reproduce the computation, would need to store about 10x more state.

File: torch/distributions.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions.py]: Done.

**Comment by fritzo** [comment]
@alicanb Would you be up for reviewing the statistical aspects of this PR?

**Comment by alicanb** [comment]
Sure

File: aten/src/TH/THRandom.c [hunk 5]
```diff
@@ -305,6 +305,51 @@ double THRandom_standard_gamma(THGenerator *_generator, double alpha) {
...
+    const double e = 1.0 - v + log(v);
+    const double accept = 0.5 * x * x + d * e;
+    if(log(u) < accept) {
+      const double y_grad = c_grad * x;
+      const double v_grad = 3 * y * y * y_grad;
+      const double e_grad = (1 / v - 1) * v_grad;
+      const double accept_grad = d_grad * e + d * e_grad;
+      const double dv = d * v;
+      const double dv_grad = d_grad * v + d * v_grad      // Pathwise part.
+                           + (dv - alpha) * accept_grad;  // Acceptance part.
```
Comment (fritzo) [inline aten/src/TH/THRandom.c]: This is the trickiest part. This roughly follows [Naesseth et al. (2017)](http://proceedings.mlr.press/v54/naesseth17a/naesseth17a.pdf) but computes an exact reparameterized gradient rather than approximating. Naesseth et al. seem to be missing the analytical baseline `alpha` in `(dv - alpha)` which is easy to compute. Note that `accept = log(acceptance ratio)`, so we're using the log trick in multiplying by `accept_grad` (and avoiding an expensive `exp()`).

File: aten/src/TH/generic/THTensorRandom.c [hunk 6]
```diff
@@ -111,18 +135,58 @@ void THTensor_(exponential)(THTensor *self, THGenerator *_generator, double lamb
...
+  TH_CHECK_CONTIGUOUS(saved_u);
+  TH_CHECK_CONTIGUOUS(saved_x);
+  TH_CHECK_SAME_SIZE(saved_u, alpha);
+  TH_CHECK_SAME_SIZE(saved_x, alpha);
+
+  real*const sample_data = THTensor_(data)(sample);
+  real*const alpha_data = THTensor_(data)(alpha);
+  real*const saved_u_data = THTensor_(data)(saved_u);
+  real*const saved_x_data = THTensor_(data)(saved_x);
+  for(int64_t i = 0, numel = THTensor_(nElement)(alpha); i < numel; ++i) {
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorRandom.c]: Any reason why you don't use OpenMP here?

File: torch/distributions.py [hunk 7]
```diff
@@ -201,12 +201,16 @@ def log_prob(self, value):
...
-        return torch._C._standard_gamma_with_grad(alpha, ctx.saved_grad)
+        ctx.save_for_backward(alpha)
+        ctx.saved_u = alpha.new(alpha.size())
+        ctx.saved_x = alpha.new(alpha.size())
+        return torch._C._standard_gamma_fwd(alpha, ctx.saved_u, ctx.saved_x)
 
     @staticmethod
     def backward(ctx, grad_output):
-        return grad_output * Variable(ctx.saved_grad)
+        alpha, = ctx.saved_variables
```
Comment (apaszke) [inline torch/distributions.py]: This backward isn't really differentiable twice. Can you mark it `@once_differentiable` and use `ctx.saved_tensors`? `grad_output` will become a tensor too.

File: torch/csrc/generic/methods/TensorRandom.cwrap [hunk 8]
```diff
@@ -240,15 +240,35 @@
...
+  return: argument 0
+  variants:
+    - function
+  options:
+    - cname: standard_gamma_bwd
+      arguments:
+        - arg: THTensor* output
+          output: True
+        - THTensor* alpha
+        - THTensor* saved_u
```
Comment (apaszke) [inline torch/csrc/generic/methods/TensorRandom.cwrap]: It would be nice if we could skip these outputs if we knew an op won't be differentiated. But it's fine as is, and we can fix that later.

File: aten/src/TH/generic/THTensorRandom.c (hunk 6 above)
Comment (fritzo) [reply aten/src/TH/generic/THTensorRandom.c]: It's difficult to parallelize because `THGenerator *gen` is stateful. However this `_fwd()` step is much cheaper than the `_bwd()` step.

File: torch/csrc/generic/methods/TensorRandom.cwrap (hunk 8 above)
Comment (fritzo) [reply torch/csrc/generic/methods/TensorRandom.cwrap]: Already done :-) `standard_gamma()` is a simplified version of `standard_gamma_fwd()` where the unused outputs are discarded.

**Comment by alicanb** [comment]
Do you know the difference between `torch.set_rng_state` and `torch.manual_seed`? I tried `torch.manual_seed` while testing normal but couldn't get it working so I used `set_rng_state`

**Comment by apaszke** [comment]
They are pretty much equivalent.
```python
torch.manual_seed(2)
s = torch.get_rng_state()
print(torch.randn(1))
# Same
torch.manual_seed(2)
print(torch.randn(1)) 
```

File: test/test_distributions.py [hunk 9]
```diff
@@ -163,6 +166,48 @@ def test_gamma_sample(self):
                                         scipy.stats.gamma(alpha, scale=1 / beta),
```
Comment (alicanb) [inline test/test_distributions.py]: 1 / beta throws typeerror on python3.  1->1.0 fixes it

File: test/test_distributions.py [hunk 10]
```diff
@@ -163,6 +166,48 @@ def test_gamma_sample(self):
...
 
+    # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_gamma_sample_with_grad(self):
+        self._set_rng_seed()
+        for alpha, beta in product([0.1, 1.0, 5.0], [0.1, 1.0, 10.0]):
+            alphas = Variable(torch.Tensor([alpha]), requires_grad=True)
+            betas = Variable(torch.Tensor([beta]), requires_grad=True)
+            self._check_sampler_sampler(Gamma(alphas, betas),
+                                        scipy.stats.gamma(alpha, scale=1 / beta),
```
Comment (alicanb) [inline test/test_distributions.py]: Same here

**Comment by fritzo** [comment]
Ok, I've changed algorithms to now directly approximate the reparameterized gradient. This achieves simpler code, cheaper computation, and more accurate gradients (they are no longer stochastic for alpha < 1).

**Comment by fritzo** [comment]
@tbrx @jwvdm I've added a Jupyter notebook and some explanation in the PR description. Let me know if I can answer any other questions. Thanks for offering to review!

**Comment by jwvdm** [comment]
Thanks @fritzo: I'll plan on taking some time to review on Fri.

**Comment by fritzo** [comment]
Moving to https://github.com/pytorch/pytorch/pull/3978
